File: sample-objc/SendBird-iOS/GroupChannel/CreateGroupChannelUserListViewController.m
//
//  CreateGroupChannelUserListViewController.m
//  SendBird-iOS
//
//  Created by Jed Kyung on 9/21/16.
//  Copyright © 2016 SendBird. All rights reserved.
//

#import <SendBirdSDK/SendBirdSDK.h>

#import "CreateGroupChannelUserListViewController.h"
#import "SelectedUserListCollectionViewCell.h"
#import "CreateGroupChannelUserListTableViewCell.h"
#import "CreateGroupChannelSelectOptionViewController.h"
#import "NSBundle+SendBird.h"
#import "Constants.h"

@interface CreateGroupChannelUserListViewController ()

@property (weak, nonatomic) IBOutlet UINavigationItem *navItem;
@property (weak, nonatomic) IBOutlet UICollectionView *selectedUserListCollectionView;
@property (weak, nonatomic) IBOutlet UITableView *userListTableView;
@property (strong, nonatomic) UIRefreshControl *refreshControl;

@property (strong, nonatomic) NSMutableArray<SBDUser *> *users;
@property (strong, nonatomic) SBDUserListQuery *userListQuery;
@property (strong, nonatomic) NSMutableArray<SBDUser *> *selectedUsers;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *selectedUserListHeight;

@end

@implementation CreateGroupChannelUserListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    UIBarButtonItem *negativeLeftSpacer = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace target:nil action:nil];
    negativeLeftSpacer.width = -2;
    UIBarButtonItem *negativeRightSpacer = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace target:nil action:nil];
    negativeRightSpacer.width = -2;

    UIBarButtonItem *leftCloseItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"btn_close"] style:UIBarButtonItemStyleDone target:self action:@selector(close)];
    UIBarButtonItem *rightNextItem = [[UIBarButtonItem alloc] initWithTitle:[NSBundle sbLocalizedStringForKey:@"NextButton"] style:UIBarButtonItemStylePlain target:self action:@selector(next)];
    [rightNextItem setTitleTextAttributes:@{NSFontAttributeName: [Constants navigationBarButtonItemFont]} forState:UIControlStateNormal];
    
    self.navItem.leftBarButtonItems = @[negativeLeftSpacer, leftCloseItem];
    self.navItem.rightBarButtonItems = @[negativeRightSpacer, rightNextItem];

    self.selectedUserListCollectionView.contentInset = UIEdgeInsetsMake(0, 14, 0, 14);
    self.selectedUserListCollectionView.delegate = self;
    self.selectedUserListCollectionView.dataSource = self;
    [self.selectedUserListCollectionView registerNib:[SelectedUserListCollectionViewCell nib] forCellWithReuseIdentifier:[SelectedUserListCollectionViewCell cellReuseIdentifier]];
    self.selectedUserListHeight.constant = 0;
    self.selectedUserListCollectionView.hidden = YES;
    
    self.userListTableView.delegate = self;
    self.userListTableView.dataSource = self;
    [self.userListTableView registerNib:[CreateGroupChannelUserListTableViewCell nib] forCellReuseIdentifier:[CreateGroupChannelUserListTableViewCell cellReuseIdentifier]];
    
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.refreshControl addTarget:self action:@selector(refreshUserList) forControlEvents:UIControlEventValueChanged];
    [self.userListTableView addSubview:self.refreshControl];
    
    self.selectedUsers = [[NSMutableArray alloc] init];

    [self.view layoutIfNeeded];
    
    [self loadUserList:YES];
}

- (void)close {
    [self dismissViewControllerAnimated:NO completion:nil];
}

- (void)next {
    if (self.selectedUsers.count == 0) {
        return;
    }
    
    CreateGroupChannelSelectOptionViewController *vc = [[CreateGroupChannelSelectOptionViewController alloc] init];
    vc.selectedUser = [[NSArray alloc] initWithArray:self.selectedUsers];
    vc.delegate = self;
    [self presentViewController:vc animated:NO completion:nil];
}

- (void)refreshUserList {
    [self loadUserList:YES];
}

- (void)loadUserList:(BOOL)initial {
    if (initial == YES) {
        if (self.users == nil) {
            self.users = [[NSMutableArray alloc] init];
        }
        
        if (self.selectedUsers == nil) {
            self.selectedUsers = [[NSMutableArray alloc] init];
        }
        
        [self.selectedUsers removeAllObjects];
        [self.users removeAllObjects];
        dispatch_async(dispatch_get_main_queue(), ^{
            self.selectedUserListHeight.constant = 0;
            self.selectedUserListCollectionView.hidden = YES;
            
            [self.userListTableView reloadData];
            [self.selectedUserListCollectionView reloadData];
        });
        
        self.userListQuery = nil;
    }
    
    if (self.userListQuery == nil) {
        self.userListQuery = [SBDMain createAllUserListQuery];
        self.userListQuery.limit = 25;
    }
    
    if (self.userListQuery.hasNext == NO) {
        [self.refreshControl endRefreshing];
        return;
    }

    [self.userListQuery loadNextPageWithCompletionHandler:^(NSArray<SBDUser *> * _Nullable users, SBDError * _Nullable error) {
        if (error != nil) {
            UIAlertController *vc = [UIAlertController alertControllerWithTitle:[NSBundle sbLocalizedStringForKey:@"ErrorTitle"] message:error.domain preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction *closeAction = [UIAlertAction actionWithTitle:[NSBundle sbLocalizedStringForKey:@"CloseButton"] style:UIAlertActionStyleCancel handler:nil];
            [vc addAction:closeAction];
            dispatch_async(dispatch_get_main_queue(), ^{
                [self presentViewController:vc animated:YES completion:nil];
            });
            
            [self.refreshControl endRefreshing];
            
            return;
        }
        
        for (SBDUser *user in users) {
            if ([user.userId isEqualToString:[SBDMain getCurrentUser].userId]) {
                continue;
            }
            [self.users addObject:user];
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.userListTableView reloadData];
        });
        
        [self.refreshControl endRefreshing];
    }];
}

#pragma mark - UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.selectedUsers.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    SelectedUserListCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:[SelectedUserListCollectionViewCell cellReuseIdentifier] forIndexPath:indexPath];
    [cell setModel:self.selectedUsers[indexPath.row]];
    
    return cell;
}

#pragma mark - UICollectionViewDelegate
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    SBDUser *selectedUser = self.selectedUsers[indexPath.row];
    [self.selectedUsers removeObject:selectedUser];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.selectedUsers.count == 0) {
            self.selectedUserListHeight.constant = 0;
            self.selectedUserListCollectionView.hidden = YES;
        }
        [collectionView reloadData];
        [self.userListTableView reloadData];
    });
}

#pragma mark - UITableViewDelegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 56;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 56;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForHeaderInSection:(NSInteger)section
{
    return 0;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForFooterInSection:(NSInteger)section
{
    return 0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    SBDUser *selectedUser = self.users[indexPath.row];
    
    if ([self.selectedUsers indexOfObject:selectedUser] == NSNotFound) {
        [self.selectedUsers addObject:selectedUser];
    }
    else {
        [self.selectedUsers removeObject:selectedUser];
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.selectedUsers.count > 0) {
            self.selectedUserListHeight.constant = 90;
            self.selectedUserListCollectionView.hidden = NO;
        }
        else {
            self.selectedUserListHeight.constant = 0;
            self.selectedUserListCollectionView.hidden = YES;
        }
        
        [self.userListTableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationNone];
        [self.selectedUserListCollectionView reloadData];
    });
}

#pragma mark - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.users.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    CreateGroupChannelUserListTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:[CreateGroupChannelUserListTableViewCell cellReuseIdentifier]];
    [cell setModel:self.users[indexPath.row]];
    
    if ([self.selectedUsers indexOfObject:self.users[indexPath.row]] == NSNotFound) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [cell setSelectedUser:NO];
        });
    }
    else {
        dispatch_async(dispatch_get_main_queue(), ^{
            [cell setSelectedUser:YES];
        });
    }
    
    if (self.users.count > 0 && indexPath.row + 1 == self.users.count) {
        [self loadUserList:NO];
    }
    
    return cell;
}

#pragma mark - CreateGroupChannelSelectOptionViewControllerDelegate
- (void)didFinishCreatingGroupChannel:(SBDGroupChannel *)channel viewController:(UIViewController *)vc {
    [self dismissViewControllerAnimated:NO completion:^{
        if (self.delegate != nil) {
            [self.delegate openGroupChannel:channel viewController:self];
        }
    }];
}

@end

File: sample-swift/SendBird-iOS/GroupChannel/CreateGroupChannelUserListViewController.swift
//
//  CreateGroupChannelUserListViewController.swift
//  SendBird-iOS
//
//  Created by Jed Kyung on 10/10/16.
//  Copyright © 2016 SendBird. All rights reserved.
//

import UIKit
import SendBirdSDK

protocol CreateGroupChannelUserListViewControllerDelegate: class {
    func openGroupChannel(channel: SBDGroupChannel, vc: UIViewController)
}

class CreateGroupChannelUserListViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate, UITableViewDelegate, UITableViewDataSource, CreateGroupChannelSelectOptionViewControllerDelegate {
    weak var delegate: CreateGroupChannelUserListViewControllerDelegate?
    var userSelectionMode: Int = 0
    var groupChannel: SBDGroupChannel?
    
    @IBOutlet weak var navItem: UINavigationItem!
    @IBOutlet weak var selectedUserListCollectionView: UICollectionView!
    @IBOutlet weak var userListTableView: UITableView!
    @IBOutlet weak var selectedUserListHeight: NSLayoutConstraint!
    
    private var refreshControl: UIRefreshControl!
    private var users: [SBDUser] = []
    private var userListQuery: SBDUserListQuery?
    private var selectedUsers: [SBDUser] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        let negativeLeftSpacer = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.fixedSpace, target: nil, action: nil)
        negativeLeftSpacer.width = -2
        let negativeRightSpacer = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.fixedSpace, target: nil, action: nil)
        negativeRightSpacer.width = -2
        
        let leftCloseItem = UIBarButtonItem(image: UIImage(named: "btn_close"), style: UIBarButtonItemStyle.done, target: self, action: #selector(close))
        var rightNextItem: UIBarButtonItem?
        if self.userSelectionMode == 0 {
            rightNextItem = UIBarButtonItem(title: Bundle.sbLocalizedStringForKey(key: "NextButton"), style: UIBarButtonItemStyle.plain, target: self, action: #selector(nextStage))
            rightNextItem?.setTitleTextAttributes([NSAttributedStringKey.font : Constants.navigationBarButtonItemFont()], for: UIControlState.normal)
        }
        else {
            rightNextItem = UIBarButtonItem(title: Bundle.sbLocalizedStringForKey(key: "InviteButton"), style: UIBarButtonItemStyle.plain, target: self, action: #selector(invite))
            rightNextItem?.setTitleTextAttributes([NSAttributedStringKey.font : Constants.navigationBarButtonItemFont()], for: UIControlState.normal)
        }
        
        self.navItem.leftBarButtonItems = [negativeLeftSpacer, leftCloseItem]
        self.navItem.rightBarButtonItems = [negativeRightSpacer, rightNextItem!]
        
        self.selectedUserListCollectionView.contentInset = UIEdgeInsetsMake(0, 14, 0, 14)
        self.selectedUserListCollectionView.delegate = self
        self.selectedUserListCollectionView.dataSource = self
        self.selectedUserListCollectionView.register(SelectedUserListCollectionViewCell.nib(), forCellWithReuseIdentifier: SelectedUserListCollectionViewCell.cellReuseIdentifier())
        self.selectedUserListHeight.constant = 0
        self.selectedUserListCollectionView.isHidden = true
        
        self.userListTableView.delegate = self
        self.userListTableView.dataSource = self
        self.userListTableView.register(CreateGroupChannelUserListTableViewCell.nib(), forCellReuseIdentifier: CreateGroupChannelUserListTableViewCell.cellReuseIdentifier())
        
        self.refreshControl = UIRefreshControl()
        self.refreshControl.addTarget(self, action: #selector(refreshUserList), for: UIControlEvents.valueChanged)
        self.userListTableView.addSubview(self.refreshControl)
        
        self.view.layoutIfNeeded()
        
        self.loadUserList(initial: true)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @objc private func close() {
        self.dismiss(animated: false, completion: nil)
    }
    
    @objc private func nextStage() {
        if self.selectedUsers.count == 0 {
            return
        }
        
        let vc = CreateGroupChannelSelectOptionViewController(nibName: "CreateGroupChannelSelectOptionViewController", bundle: Bundle.main)
        vc.selectedUser = NSArray(array: self.selectedUsers) as! [SBDUser]
        vc.delegate = self
        self.present(vc, animated: false, completion: nil)
    }
    
    @objc private func invite() {
        if self.selectedUsers.count == 0 {
            return
        }
        
        self.groupChannel?.invite(self.selectedUsers, completionHandler: { (error) in
            DispatchQueue.main.async {
                self.dismiss(animated: false, completion: nil)
            }
        })
    }
    
    @objc private func refreshUserList() {
        self.loadUserList(initial: true)
    }
    
    @objc private func loadUserList(initial: Bool) {
        if initial == true {
            self.users.removeAll()
            self.selectedUsers.removeAll()
            
            DispatchQueue.main.async {
                self.selectedUserListHeight.constant = 0
                self.selectedUserListCollectionView.isHidden = true
                
                self.userListTableView.reloadData()
                self.selectedUserListCollectionView.reloadData()
            }
            
            self.userListQuery = nil;
        }
        
        if self.userListQuery == nil {
            self.userListQuery = SBDMain.createAllUserListQuery()
            self.userListQuery?.limit = 25
        }
        
        if self.userListQuery?.hasNext == false {
            self.refreshControl.endRefreshing()
            return
        }
        
        self.userListQuery?.loadNextPage(completionHandler: { (users, error) in
            if error != nil {
                let vc = UIAlertController(title: Bundle.sbLocalizedStringForKey(key: "ErrorTitle"), message: error?.domain, preferredStyle: UIAlertControllerStyle.alert)
                let closeAction = UIAlertAction(title: Bundle.sbLocalizedStringForKey(key: "CloseButton"), style: UIAlertActionStyle.cancel, handler:nil)
                vc.addAction(closeAction)
                DispatchQueue.main.async {
                    self.present(vc, animated: true, completion: nil)
                }
                
                self.refreshControl.endRefreshing()
                
                return
            }
            
            for user in users! as [SBDUser] {
                if user.userId == SBDMain.getCurrentUser()?.userId {
                    continue
                }
                
                if self.userSelectionMode == 1 {
                    var isMember: Bool = false
                    for item in (self.groupChannel?.members)! {
                        let member: SBDUser = item as! SBDUser
                        if member.userId == user.userId {
                            isMember = true
                            break
                        }
                    }
                    
                    if isMember == false {
                        self.users.append(user)
                    }
                }
                else {
                    self.users.append(user)
                }
            }
            
            DispatchQueue.main.async {
                self.userListTableView.reloadData()
            }
            
            self.refreshControl.endRefreshing()
        })
    }
    
    // MARK: UICollectionViewDataSource
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.selectedUsers.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = (collectionView.dequeueReusableCell(withReuseIdentifier: SelectedUserListCollectionViewCell.cellReuseIdentifier(), for: indexPath)) as! SelectedUserListCollectionViewCell
        cell.setModel(aUser: self.selectedUsers[indexPath.row])
        
        return cell
    }
    
    // MARK: UICollectionViewDelegate
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        self.selectedUsers.remove(at: indexPath.row)
        
        DispatchQueue.main.async {
            if self.selectedUsers.count == 0 {
                self.selectedUserListHeight.constant = 0
                self.selectedUserListCollectionView.isHidden = true
            }
            collectionView.reloadData()
            self.userListTableView.reloadData()
        }
    }
    
    // MARK: UITableViewDelegate
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 56
    }
    
    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return 56
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 0
    }
    
    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 0
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
        let selectedUser = self.users[indexPath.row] as SBDUser
        
        if self.selectedUsers.index(of: selectedUser) == nil {
            self.selectedUsers.append(selectedUser)
        }
        else {
            if self.selectedUsers.index(of: selectedUser) != nil {
                self.selectedUsers.remove(at: self.selectedUsers.index(of: selectedUser)!)
            }
        }
        
        DispatchQueue.main.async {
            if self.selectedUsers.count > 0 {
                self.selectedUserListHeight.constant = 70
                self.selectedUserListCollectionView.isHidden = false
            }
            else {
                self.selectedUserListHeight.constant = 0
                self.selectedUserListCollectionView.isHidden = true
            }
            
            self.userListTableView.reloadRows(at: [indexPath], with: UITableViewRowAnimation.none)
            self.selectedUserListCollectionView.reloadData()
        }
    }
    
    // MARK: UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.users.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: CreateGroupChannelUserListTableViewCell.cellReuseIdentifier()) as! CreateGroupChannelUserListTableViewCell
        cell.setModel(aUser: self.users[indexPath.row])
        
        if self.selectedUsers.index(of: self.users[indexPath.row]) == nil {
            DispatchQueue.main.async {
                cell.setSelectedUser(selected: false)
            }
        }
        else {
            DispatchQueue.main.async {
                cell.setSelectedUser(selected: true)
            }
        }
        
        if self.users.count > 0 && indexPath.row + 1 == self.users.count {
            self.loadUserList(initial: false)
        }
        
        return cell
    }
    
    // MARK: CreateGroupChannelSelectOptionViewControllerDelegate
    func didFinishCreating(channel: SBDGroupChannel, vc: UIViewController) {
        self.dismiss(animated: false) { 
            if self.delegate != nil {
                self.delegate?.openGroupChannel(channel: channel, vc: self)
            }
        }
    }
}
